### PR TITLE
Fixing migrations for CompetencyAssessmentQuestionRoleRequirements in…

### PR DESCRIPTION
**Pull request by DanBloxham-sw**
… HEEDLS-741

### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-741

### Description
Created two new migrations, one to undo the previous constraint-introducing migrations (depending on the state the DB is in), and one to add the new unique constraint.
I also commented out the previous migrations to prevent them from causing more errors when migrating past them. The "Undo" migration takes this into account.
I also modified one of the automated test files, as the set up of a fake server was causing errors when it tried to migrate up as it didn't have a config file to take a connection string from. I suspect it could probably be made more efficient in future.

### Screenshots
N/A

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [ ] Run the formatter and made sure there are no IDE errors.
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data.Migrations/202201171115_UndoPreviousConstraintMigrationsForCompetencyAssessmentQuestionRoleRequirements.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    [Migration(202201171115)]
+    public class UndoPreviousConstraintMigrationsForCompetencyAssessmentQuestionRoleRequirements : Migration
+    {
+        public override void Up()
+        {
+            if (Schema.Table("CompetencyAssessmentQuestionRoleRequirements").Constraint(
+                "IX_CompetencyAssessmentQuestionRoleRequirements_SelfAssessmentID_CompetencyID"
+            ).Exists())
+            {
+                // This undoes the changes from 202201120821_ChangeUniqueConstraintsOnCompetencyAssessmentQuestionRoleRequirements
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Migrations/202201171115_UndoPreviousConstraintMigrationsForCompetencyAssessmentQuestionRoleRequirements.cs:14]: I'll be picky and suggest instead "This undoes the **unwanted** changes..."

File: DigitalLearningSolutions.Data.Migrations/202217011130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+﻿namespace DigitalLearningSolutions.Data.Migrations
+{
+    using FluentMigrator;
+
+    [Migration(202217011130)]
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Migrations/202217011130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs:5]: Date looks wrong - it's got year day month...

Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Migrations/202217011130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs:5]: Other migration is fine.

**Review by SteveJacksonSoft — CHANGES_REQUESTED** [review]

File: DigitalLearningSolutions.Data.Migrations/202201171115_UndoPreviousConstraintMigrationsForCompetencyAssessmentQuestionRoleRequirements.cs (hunk 1 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data.Migrations/202201171115_UndoPreviousConstraintMigrationsForCompetencyAssessmentQuestionRoleRequirements.cs]: That's an appropriate suggestion, I've added it.

File: DigitalLearningSolutions.Data.Migrations/202217011130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs (hunk 2 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Data.Migrations/202217011130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs]: Good spot, I've fixed the ID and file name.

File: DigitalLearningSolutions.Data.Migrations/202201171130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿namespace DigitalLearningSolutions.Data.Migrations
+{
+    using FluentMigrator;
+
+    [Migration(202201171130)]
+    public class AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements : Migration
+    {
+        public override void Up()
+        {
+            Create.UniqueConstraint(
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Data.Migrations/202201171130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs:10]: Is it worth adding a DELETE query in here to remove any existing records that won't fit the new constraint? i.e. any CompetencyAssessmentQuestionRoleRequirements that aren't the first record with the same SelfAssessmentID,                CompetencyID, AssessmentQuestionID, LevelValue combination - just in case?

File: DigitalLearningSolutions.Web.AutomatedUiTests/SeleniumServerFactory.cs [hunk 4]
```diff
@@ -31,7 +31,11 @@ protected sealed override TestServer CreateServer(IWebHostBuilder builder)
             RootUri = host.ServerFeatures.Get<IServerAddressesFeature>().Addresses.First();
 
             // Fake Server to satisfy the return type
-            return new TestServer(new WebHostBuilder().UseStartup<TStartup>().UseSerilog());
+            return new TestServer(new WebHostBuilder()
```
Comment (kevwhitt-hee) [inline DigitalLearningSolutions.Web.AutomatedUiTests/SeleniumServerFactory.cs:34]: Is this change related to this PR?

Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Web.AutomatedUiTests/SeleniumServerFactory.cs:34]: I see that it is. Ignore.

File: DigitalLearningSolutions.Data.Migrations/202201171130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs (hunk 3 above)
Comment (kevwhitt-hee) [reply DigitalLearningSolutions.Data.Migrations/202201171130_AddUniqueConstraintToCompetencyAssessmentQuestionRoleRequirements.cs:10]: Discussed this with Steve. Ignore this too!

**Review by kevwhitt-hee — APPROVED** [review]
Happy with this once Steve's requested changes are in place.

**Review by SteveJacksonSoft — APPROVED** [review]

**Comment by DanBloxham-sw** [comment]
Created new branch for this. See PR: https://github.com/TechnologyEnhancedLearning/DLSV2/pull/866
